components/camera/replaypcd/replaypcd_utils_test.go

```diff
@@ -0,0 +1,207 @@
+	"path/filepath"
+	"strconv"
+	"testing"
+
+	"github.com/edaniels/golog"
+	datapb "go.viam.com/api/app/data/v1"
+	"go.viam.com/test"
+	"go.viam.com/utils/artifact"
+	"go.viam.com/utils/rpc"
+
+	"go.viam.com/rdk/components/camera"
+	viamgrpc "go.viam.com/rdk/grpc"
+	"go.viam.com/rdk/internal/cloud"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/robot"
+	"go.viam.com/rdk/testutils/inject"
+)
+
+// mockDataServiceServer is a struct that includes unimplemented versions of all the Data Service endpoints. These
+// can be overwritten to allow developers to triggers desired behaviors during testing.
+type mockDataServiceServer struct {
+	datapb.UnimplementedDataServiceServer
+}
+
+// BinaryDataByFilter is a mocked version of the Data Service function of a similar name. It returns a response with
+// data corresponding to a stored pcd artifact based on the filter and last file accessed.
+func (mDServer *mockDataServiceServer) BinaryDataByFilter(ctx context.Context, req *datapb.BinaryDataByFilterRequest,
+) (*datapb.BinaryDataByFilterResponse, error) {
+	// Parse request
+	filter := req.DataRequest.GetFilter()
+	last := req.DataRequest.GetLast()
+
+	newFileNum, err := getNextDataAfterFilter(filter, last)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get point cloud data in gzip compressed format
+	path := filepath.Clean(artifact.MustPath(fmt.Sprintf(datasetDirectory, newFileNum)))
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var dataBuf bytes.Buffer
+	gz := gzip.NewWriter(&dataBuf)
+	gz.Write(data)
+	gz.Close()
+
+	// Construct response
+	binaryData := &datapb.BinaryData{
+		Binary:   dataBuf.Bytes(),
+		Metadata: &datapb.BinaryMetadata{},
+	}
+
+	resp := &datapb.BinaryDataByFilterResponse{
+		Data: []*datapb.BinaryData{binaryData},
+		Last: fmt.Sprint(newFileNum),
+	}
+	return resp, nil
+}
+
+// createMockCloudDependencies creates a mockDataServiceServer and rpc client connection to it which is then
+// stored in a mockCloudConnectionService.
+func createMockCloudDependencies(ctx context.Context, t *testing.T, logger golog.Logger) (resource.Dependencies, func() error) {
+	listener, err := net.Listen("tcp", "localhost:0")
+	test.That(t, err, test.ShouldBeNil)
+	rpcServer, err := rpc.NewServer(logger, rpc.WithUnauthenticated())
+	test.That(t, err, test.ShouldBeNil)
+
+	test.That(t, rpcServer.RegisterServiceServer(
+		ctx,
+		&datapb.DataService_ServiceDesc,
+		&mockDataServiceServer{},
+		datapb.RegisterDataServiceHandlerFromEndpoint,
+	), test.ShouldBeNil)
+
+	go rpcServer.Serve(listener)
+
+	conn, err := viamgrpc.Dial(ctx, listener.Addr().String(), logger)
+	test.That(t, err, test.ShouldBeNil)
+
+	r := &inject.Robot{}
+	rs := map[resource.Name]resource.Resource{}
+	rs[cloud.InternalServiceName] = &mockCloudConnectionService{
+		Named: cloud.InternalServiceName.AsNamed(),
+		conn:  conn,
+	}
+
+	r.MockResourcesFromMap(rs)
+
+	return resourcesFromDeps(t, r, []string{cloud.InternalServiceName.String()}), rpcServer.Stop
```

Comment:
you'll need to close the conn too so just wrap this in a bigger `func() error { ... }`

Reply:
I was thinking about doing just that, however, in the function `closeCloudConnection` in `replaypcd.go`, I close the provided cloudConn. And since. I call Close at the end of every run, this should get closed there.